### PR TITLE
fix flags for tracing & warnings

### PR DESCRIPTION
**Pull request by conorbros**
Finish implementing flags and their appropriate handling in the codec methods

### CODE REVIEW

**Comment by rukai** [comment]
What about moving Flags::CUSTOM_PAYLOAD and Flags::BETA to be bool fields on the frame and then generating the flags at encode time?

As is, this PR has an important fix though, so lets not block on that.

File: cassandra-protocol/src/frame.rs [hunk 1]
```diff
@@ -255,18 +255,54 @@ impl Envelope {
         v.extend_from_slice(&self.stream_id.to_be_bytes());
         v.push(opcode_byte);
 
+        let mut body_buffer = vec![];
+
+        if self.flags.contains(Flags::TRACING) && self.direction == Direction::Response {
+            let mut tracing_id = self
+                .tracing_id
+                .expect("Tracing flag was set but Envelope has no tracing_id")
```
Comment (krojew) [inline cassandra-protocol/src/frame.rs:263]: We cannot crash on invalid data.

**Review by krojew — CHANGES_REQUESTED** [review]
This change will break compression for v4 - the whole envelope body is expected to be compressed, rather than consisting of several compressed parts.

File: cassandra-protocol/src/frame.rs [hunk 2]
```diff
@@ -255,16 +255,48 @@ impl Envelope {
...
+
+            let warnings_len = self.warnings.len() as i16;
+            body_buffer.extend_from_slice(&warnings_len.to_be_bytes());
+            body_buffer.append(&mut warnings_buffer);
+        }
+
+        body_buffer.extend_from_slice(&self.body);
 
+        if is_compressed {
+            let encoded_body = compressor.encode(&body_buffer)?;
```
Comment (krojew) [inline cassandra-protocol/src/frame.rs:292]: We can reuse body_buffer here to avoid eating twice the memory.

**Review by krojew — CHANGES_REQUESTED** [review]
Generally, looks good, but I'm concerned about the number of expensive allocations in such critical part. The typical case is encoding envelopes without tracing or warnings. In such situations the current implementation copies the whole envelope to a buffer as is, which is then copied to the output with optional compression. We have an unnecessary allocation + copy which can hurt performance.

**Comment by conorbros** [comment]
Most recent commit addresses those concerns. In the typical case (no flags) we can avoid having to concatenate the buffers.

File: cassandra-protocol/src/frame.rs [hunk 3]
```diff
@@ -255,16 +255,59 @@ impl Envelope {
...
+
+            flags_buffer.append(&mut tracing_id);
+        };
+
+        if self.flags.contains(Flags::WARNING) && self.direction == Direction::Response {
+            let mut warnings_buffer = vec![];
+
+            for warning in &self.warnings {
+                let warning_len = warning.len() as i16;
+                warnings_buffer.extend_from_slice(&warning_len.to_be_bytes());
```
Comment (krojew) [inline cassandra-protocol/src/frame.rs:280]: You can append to flags_buffer directly, thus avoiding a temporary buffer.

File: cassandra-protocol/src/frame.rs [hunk 4]
```diff
@@ -255,16 +255,59 @@ impl Envelope {
...
+            flags_buffer.append(&mut warnings_buffer);
+        }
+
         if is_compressed {
-            let mut encoded_body = compressor.encode(&self.body)?;
+            // avoid concating the slices if there is nothing in flags_buffer
+            let encoded_body = if flags_buffer.is_empty() {
+                compressor.encode(&self.body)?
+            } else {
+                compressor.encode(&[flags_buffer.as_slice(), &self.body].concat())?
```
Comment (krojew) [inline cassandra-protocol/src/frame.rs:294]: `concat` should not be used as it creates a temporary vector and copies all the slices into it.

File: cassandra-protocol/src/frame.rs [hunk 5]
```diff
@@ -255,16 +255,59 @@ impl Envelope {
...
-            v.extend_from_slice(&self.body);
+            // avoid concating the slices if there is nothing in flags_buffer
+            if flags_buffer.is_empty() {
+                let body_len = self.body.len() as i32;
+                v.extend_from_slice(&body_len.to_be_bytes());
+                v.extend_from_slice(&self.body);
+            } else {
+                let body_len = self.body.len() as i32 + flags_buffer.len() as i32;
+                v.extend_from_slice(&body_len.to_be_bytes());
+                v.extend_from_slice(&[flags_buffer.as_slice(), &self.body].concat());
```
Comment (krojew) [inline cassandra-protocol/src/frame.rs:309]: As above.

**Review by krojew — CHANGES_REQUESTED** [review]

**Comment by krojew** [comment]
Thanks!

**Pull request merged**